### PR TITLE
APP-3287 SlashAnnotationProcess scans singleton beans only

### PR DESCRIPTION
**Pull request by symphony-thibault**
### Issue
https://github.com/SymphonyPlatformSolutions/symphony-api-client-java/issues/345

### Description
`SlashAnnotationProcessor` should only scan beans with scope "singleton". 

### Dependencies
N/A

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Javadoc added or updated
- [x] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/config/BdkActivityConfig.java [hunk 1]
```diff
@@ -88,7 +90,17 @@ public void setApplicationContext(final ApplicationContext applicationContext) t
...
-      }
+      });
+    }
+
+    /**
+     * Returns a {@link Stream} of beans to be scanned from {@link ApplicationContext}.
+     * Only "singleton" beans will be returned.
+     */
+    private static Stream<Object> getBeansToScan(ApplicationContext applicationContext) {
+      return Arrays.stream(applicationContext.getBeanDefinitionNames())
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/config/BdkActivityConfig.java:101]: Don't still risk to create beans that we should not create?
Maybe lazy beans? 

I was trying to look at something similar (the application event) and the logic seems quite complex (org.springframework.context.event.EventListenerMethodProcessor#afterSingletonsInstantiated)

Comment (symphony-youri) [reply symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/config/BdkActivityConfig.java:101]: Maybe a BeanPostProcessor could be another way to do it? https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-factory-extension-bpp

Comment (symphony-thibault) [reply symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/config/BdkActivityConfig.java]: @symphony-youri see latest commit https://github.com/SymphonyPlatformSolutions/symphony-api-client-java/pull/347/commits/8b5fd0c18f9f5a2e1c1a6e446de7f42333037b52. I extracted the whole logic in a separate class, including what was done by the Spring `EventListenerMethodProcessor` class,  and based on the `BeanPostProcessor` interface as well that makes easier to handle lazy beans initialization.

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  @Override
+  public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+
+    if (!ScopedProxyUtils.isScopedTarget(beanName)) {
+      Class<?> type = null;
+      try {
+        type = AutoProxyUtils.determineTargetClass(this.applicationContext.getBeanFactory(), beanName);
+      }
+      catch (Throwable ex) {
+        // An unresolvable bean type, probably from a lazy bean - let's ignore it.
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:60]: is it the reason? I thought here the bean is already instanciated? can't we use bean.getClass() to retrieve it by the way?

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  private static boolean isMethodPrototypeValid(Method m) {
+    return m.getParameterCount() == 1 && m.getParameters()[0].getType().equals(CommandContext.class);
+  }
+
+  /**
+   * Determine whether the given class is an {@code org.springframework}
+   * bean class that is not annotated as a user or test {@link Component}...
+   * which indicates that there is no {@link Slash} to be found there.
+   */
+  private static boolean isSpringContainerClass(Class<?> clazz) {
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:143]: is that even possible to have a spring class annotated with @ Slash ?

Comment (symphony-thibault) [reply symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:143]: Called as `!isSpringContainerClass(...)` earlier in the class. Maybe I should rename and invert to `isNotSpringContainerClass(...)`

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+      final Class<?> type = this.determineTargetClass(beanName);
+
+      if (type != null) {
+        try {
+          this.processBean(bean, beanName, type);
+        }
+        catch (Throwable ex) {
+          // just alert the developer
+          log.warn("Failed to process @Slash annotation on bean with name '" + beanName + "'", ex);
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:73]: {} placeholders are usually used (for performance mainly here it won't be a problem but to be consistent)

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        }
+      }
+    }
+
+    return bean;
+  }
+
+  @Generated // means ignored from test coverage
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:81]: isn't covered by the tests?

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      log.debug("Could not resolve target class for bean with name '" + beanName + "'", ex);
+    }
+
+    if (type != null && ScopedObject.class.isAssignableFrom(type)) {
+      try {
+        Class<?> targetClass = AutoProxyUtils.determineTargetClass(this.applicationContext.getBeanFactory(), ScopedProxyUtils.getTargetBeanName(beanName));
+        if (targetClass != null) {
+          type = targetClass;
+        }
+      }
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:98]: formatting? } catch on the same line?

**Review by symphony-youri — APPROVED** [review]

**Pull request merged**